### DISCUSSION TITLE
Issue with `snprintf` handling `long long` variable incorrectly

### DISCUSSION THREAD

**Discussion by mostafa-salmaan** [picolibc/picolibc#785] [Q&A]
Hello everyone,

I'd like to start a discussion about the behavior of `snprintf` in Picolibc's tinystdio when passing a `long long` variable to it.

**Description:**
I've encountered an issue with the `snprintf` function when passing a `long long` variable to it. The function seems to handle the variable incorrectly under certain conditions. Here is a code example that demonstrates the problem:

```c
#include <stdio.h>

int main()
{
    long long unitialized_var;
    int i = 123;
    char buf[50];
    
    snprintf(buf, sizeof(buf), "%d%lln456", i, &unitialized_var);
    
    if (unitialized_var == 3)
    {
        printf("PASSED\n");
    }
    else
    {
        printf("FAILED\n");
    }
    
    return 0;
}

```

**Expected Behavior:**
The code should print "PASSED" if `unitialized_var` is correctly set to the value 3 by `snprintf`.

**Observed Behavior:**
The code prints "FAILED" when the `long long` variable is not initialized or initialized with a large value (out of the range of long), such as:

```c
long long unitialized_var = 100656853638960;
```

**Additional Information:**
- This behavior is observed only when building the Picolibc library with `-O0` optimization level.
- Initializing the `long long` variable seems to influence the outcome, which suggests that there might be an issue with how `snprintf` is handling `long long` variables.

**Discussion:**
It would be helpful to understand if this is a known issue or if there is a workaround. Are there any configurations or specific settings in Picolibc that can handle this correctly? 

Any insights or suggestions would be appreciated!

**Comment by keith-packard** [comment]
I was just noticing that the %n support in printf doesn't take the size of the parameter into account. Reading the C17 specification, we find:

> If the conversion specification includes any flags, a field with, or a precision, the behavior is undefined.

POSIX doesn't explicitly mention any extensions, so, the current picolibc implementation is conforming. You cannot use `%lln` in a printf format specifier.

**Reply by mostafa-salmaan** [reply]
Thank you for the insight on the `%n` support in printf and its limitations. I understand from your response that `%lln` is not a valid format specifier in the context of printf as per the C17 standard. and POSIX doesn't explicitly mention any extensions.

However, I’m still unclear about a couple of points:

1. Impact of Initialization: The behavior changes when the `long long` variable is initialized with a value in the range of `int`. 
Could you provide more context on why this might be affecting the outcome? Is there any specific reason why initializing the variable would influence the result if `%lln` is not supported?

2. Is the attached test not valid due to the `%lln` specifier?
so, we can't test the `long long` variable with the `%n` specifier because this behavior is undefined?

**Reply by keith-packard** [reply]
>     1. Impact of Initialization: The behavior changes when the `long long` variable is initialized with a value in the range of `int`.
>        Could you provide more context on why this might be affecting the outcome? Is there any specific reason why initializing the variable would influence the result if `%lln` is not supported?

printf is ignoring the type information in this request, so `%lln` operates exactly the same as `%n`; picolibc assumes you have passed a pointer to int and stores an int value at that address. On a little-endian machine, that ends up writing an int value to the low-order bytes of the long long variable. If that variable was initialized so that the upper portions were set to zero, the combination of the initialized data and the partial write of the int value yields the expected result.

>     2. Is the attached test not valid due to the `%lln` specifier?
>        so, we can't test the `long long` variable with the `%n` specifier because this behavior is undefined?

Correct. Printf is not specified to allow `%lln` at all, so you may not use a 'long long' variable with that format specifier. The only valid use of `%n` is with an int variable. You cannot test any of the other type modifiers, including hh, h, or l as the behavior of printf is undefined if you include those.

**Reply by mostafa-salmaan** [reply]
> I was just noticing that the %n support in printf doesn't take the size of the parameter into account. Reading the C17 specification

the C17 Specification does indeed state that 

> If the conversion specification includes any flags, a field width, or a precision, the behavior is undefined.

However, `ll` which is used in the attached test (in the form of `%lln`), is a length modifier and not any of the undefined elements of the conversion specification.

In fact, the C17 specification clearly states that a following `n` conversion specifier is valid use for each of those length modifiers.

For example, from Clause 7.21.6.1 (paragraph 7) of the C17 specification:

> The length modifiers and their meanings are:
> ll (ell-ell) Specifies that a following d, i, o, u, x, or X conversion specifier applies to a
> long long int or unsigned long long int argument; or that a following n conversion specifier applies to a pointer to a long long int argument.


As a result, I think the attached test is valid as it conforms to the specification.

**Reply by keith-packard** [reply]
Yup, dunno how I missed that.
